core/src/main/scala/frameless/InjectionEnum.scala

```diff
@@ -0,0 +1,54 @@
+package frameless
+
+import scala.reflect.{ClassTag, classTag}
+import shapeless._
+
+object InjectionEnum {
+  type InjectionEnum[A] = Injection[A, String]
+
+  def apply[A: InjectionEnum]: InjectionEnum[A] = implicitly[InjectionEnum[A]]
+
+  def instance[A](f: A => String, g: String => A): InjectionEnum[A] =
+    Injection(f, g)
+
+  implicit val cnilInjectionEnum: InjectionEnum[CNil] =
+    instance(
+      _ => throw new Exception("Impossible"),
+      name =>
+        throw new IllegalArgumentException(
+          s"Cannot construct a value of type CNil: $name did not match data constructor names"
+        )
+    )
+
+  implicit def coproductInjectionEnum[H: ClassTag, T <: Coproduct](
+    implicit
+    gen: Generic.Aux[H, HNil],
+    tInjectionEnum: InjectionEnum[T]
+    ): InjectionEnum[H :+: T] = {
+    val canonicalName = classTag[H].runtimeClass.getCanonicalName
```

Comment:
I would also like to have a test to see what this produces. Maybe using `shapeless.Typeable` would appropriate here?